proparse/src/main/java/org/prorefactor/proparse/antlr4/TokenList.java

```diff
@@ -50,99 +48,105 @@
     }
-    LOGGER.trace("Exiting TokenList#build() - {} tokens", list.size());
   }
 
   private void reviewObjcolon() {
-    int colonIndex = list.size() - 1;
-    int lastIndex = colonIndex - 1;
-
-    // Getting type of the token just before colon (excluding comments and whitespaces)
-    int ttype = list.get(lastIndex).getType();
-    while (ttype == PreprocessorParser.WS || ttype == PreprocessorParser.COMMENT) {
-      ttype = list.get(--lastIndex).getType();
+    ProToken objColonToken = queue.removeLast();
+
+    // Store comments and whitespaces before the colon
+    Deque<ProToken> comments = new LinkedList<>();
+    ProToken tok = queue.removeLast();
+    while ((tok.getType() == NodeTypes.WS) || (tok.getType() == NodeTypes.COMMENT)) {
+      comments.addFirst(tok);
+      tok = queue.pollLast();
     }
 
-    // Look for NAMEDOT pairs.
-    // Actually, it's not that easy. Something like:
-    // newsyntax.101b.deep.FindMe
-    // is perfectly valid, and because of the digit following the '.',
-    // one part of that name gets picked up as a token with text ".101b".
-    int index = lastIndex;
+    Deque<ProToken> clsName = new LinkedList<>();
     boolean foundNamedot = false;
-    for (;;) {
-      if (index == 0)
+    while (true) {
+      if (tok == null)
+        break;
+      
+      // There can be space in front of a NAMEDOT in a table or field name. We don't want to fiddle with those here.
+      if ((tok.getType() == PreprocessorParser.WS) || (tok.getType() == PreprocessorParser.COMMENT)) {
         break;
-      int currType = list.get(index).getType();
-      if (currType == PreprocessorParser.WS || currType == PreprocessorParser.COMMENT) {
-        // There can be space in front of a NAMEDOT in a table or field name.
-        // We don't want to fiddle with those here.
-        return;
       }
-      if (list.get(index - 1).getType() == PreprocessorParser.NAMEDOT) {
-        index = index - 2;
-      } else if (list.get(index).getText().charAt(0) == '.') {
-        index = index - 1;
+
+      // If previous is NAMEDOT, then we add both tokens
+      if ((queue.peekLast() != null) && (queue.peekLast().getType() == NodeTypes.NAMEDOT)) {
+        clsName.addFirst(tok);
+        clsName.addFirst(queue.pollLast());
+        tok = queue.removeLast();
+      } else if (tok.getText().startsWith(".")) {
+        clsName.addFirst(tok);
+        tok = queue.removeLast();
       } else {
         break;
       }
       foundNamedot = true;
     }
+
     if (foundNamedot) {
       // Now merge all the parts into one ID token.
-      ProToken token = list.get(index);
-      token.setType(PreprocessorParser.ID);
-      StringWriter text = new StringWriter();
-      text.append(token.getText());
-      int drop = index + 1;
-      for (int i = 0; i < lastIndex - index; i++) {
-        text.append(list.get(drop).getText());
-        list.remove(drop);
+      StringBuilder text = new StringBuilder(tok.getText());
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "tok" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS2259)